### PR TITLE
Refactor - split to modules, add parser tests, update readme

### LINKED ISSUE #196: pydocstyle fails on "raise X from Y"
The reason is that the "from" keyword triggers a "from X import Y" parsing, which fails.

```
C:\Users\rachum\code\testing_grounds>cat example.py
raise RuntimeError() from None
C:\Users\rachum\code\testing_grounds>python example.py  # this syntax is legal in Python 3
Traceback (most recent call last):
  File "example.py", line 1, in <module>
    raise RuntimeError() from None
RuntimeError

C:\Users\rachum\code\testing_grounds>pydocstyle example.py
Traceback (most recent call last):
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\rachum\AppData\Local\Programs\Python\Python35-32\Scripts\pydocstyle.exe\__main__.py", line 9, in <module>
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 1745, in main
    sys.exit(run_pydocstyle(use_pep257))
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 1360, in run_pydocstyle
    errors.extend(check((filename,), select=checked_codes))
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 1297, in check
    for error in PEP257Checker().check_source(source, filename):
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 1398, in check_source
    module = parse(StringIO(source), filename)
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 276, in __call__
    return self.parse_module()
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 429, in parse_module
    children = list(self.parse_definitions(Module, all=True))
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 376, in parse_definitions
    self.parse_from_import_statement()
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 510, in parse_from_import_statement
    is_future_import = self._parse_from_import_source()
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 525, in _parse_from_import_source
    self.check_current(value='import')
  File "c:\users\rachum\appdata\local\programs\python\python35-32\lib\site-packages\pydocstyle.py", line 501, in check_current
    assert kind_valid and value_valid, msg
AssertionError:
Unexpected token at line 2:

In file: example.py

Got kind tk.ENDMARKER
Got value
```

### PR DESCRIPTION
**Pull request by Nurdok**
@sigmavirus24 If you have some time, I'd love it if you could give this a look.
Also, this fixes #196.

### CODE REVIEW

File: requirements/tests.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 pytest==2.7.3
 pytest-pep8
 mock
-tox
\ No newline at end of file
+tox
+pathlib
```
Comment (sigmavirus24) [inline requirements/tests.txt]: So, I have a quibble with this file. Is tox a test dependency and the rest are dependencies enumerated in tox, or are these test dependencies and tox is a different kind of requirement?

File: src/main.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#! /usr/bin/env python
```
Comment (sigmavirus24) [inline src/main.py]: Is this file in the right place?

File: setup.py [hunk 3]
```diff
@@ -27,12 +29,13 @@
         'License :: OSI Approved :: MIT License',
     ],
     keywords='pydocstyle, PEP 257, pep257, PEP 8, pep8, docstrings',
+    packages=('pydocstyle',),
+    py_modules=('main',),
```
Comment (sigmavirus24) [inline setup.py]: Why are you packaging main?

File: requirements/tests.txt (hunk 1 above)
Comment (Nurdok) [inline requirements/tests.txt]: Separated to `tests.txt` and `test_env.txt`.

File: setup.py (hunk 3 above)
Comment (Nurdok) [inline setup.py]: The point of this is to allow to run `pydocstyle` from the command line without the entry point (e.g., if you want to just unzip it somewhere and run it without installing).

File: src/main.py (hunk 2 above)
Comment (Nurdok) [inline src/main.py]: See my comment in `setup.py`.

File: setup.py (hunk 3 above)
Comment (sigmavirus24) [inline setup.py]: Have you investigated how this is actually installed on a person's computer though? I think this might not do exactly what you want.

Comment (Nurdok) [inline setup.py]: I'll do some experiments to make sure.

**Comment by Nurdok** [comment]
@sigmavirus24 care to take another look? I fixed the issue with the `main` module and also had to do a manual merge from #203 .

File: src/pydocstyle/checker.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+
+def check(filenames, select=None, ignore=None):
+    """Generate PEP 257 errors that exist in `filenames` iterable.
+
+    Only returns errors with error-codes defined in `checked_codes` iterable.
+
+    Example
+    -------
+    >>> check(['pydocstyle.py'], checked_codes=['D100'])
```
Comment (sigmavirus24) [inline src/pydocstyle/checker.py]: FYI, `checked_codes` will no longer work

Comment (Nurdok) [inline src/pydocstyle/checker.py]: Updated the docstring.

File: src/pydocstyle/cli.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            errors.extend(check((filename,), select=checked_codes))
+    except IllegalConfiguration:
+        # An illegal configuration file was found during file generation.
+        return ReturnCode.invalid_options
+
+    code = ReturnCode.no_violations_found
+    count = 0
+    for error in errors:
+        sys.stderr.write('%s\n' % error)
+        code = ReturnCode.violations_found
```
Comment (sigmavirus24) [inline src/pydocstyle/cli.py]: Is it necessary to reassign this for every error? Why not do something like:

```py
count = 0
for error in errors:
    print('%s' % error)   # because we all agree sys.stderr isn't correct
    count += 1
if count > 0:
   code = ReturnCode.violations_found
else:
   code = ReturnCode.no_violations_found
```

Comment (Nurdok) [inline src/pydocstyle/cli.py]: Fixed.

**Comment by sigmavirus24** [comment]
Sorry for the lag in review @Nurdok this is a huge change, and my own personal life is a bit hectic and will be so for a couple weeks. So I'm reviewing this in chunks of free time as I find them. =)

File: src/pydocstyle/cli.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    try:
+        for filename, checked_codes in conf.get_files_to_check():
+            errors.extend(check((filename,), select=checked_codes))
+    except IllegalConfiguration:
+        # An illegal configuration file was found during file generation.
+        return ReturnCode.invalid_options
+
+    count = 0
+    for error in errors:
+        sys.stderr.write('%s\n' % error)
```
Comment (sigmavirus24) [inline src/pydocstyle/cli.py]: print instead of sys.stderr?

**Comment by sigmavirus24** [comment]
Some questions too:

- I see you have moved the tests into `src/pycodestyle`. Do you then plan to ship those tests so people can run them on their system after install? Is this for convenience? If so, what convenience?

- You have `pycodestyle/main.py` that reminds me that people can (probably, I haven't tested this) currently do `python -m pycodestyle [options] file.py` to invoke pycodestyle. As you have this structured now, they cannot do that. Do you wish to continue supporting that? (To do so, you need a `pycodestyle/__main__.py` that does pretty much exactly the same thing as `pycodestyle/main.py`.

**Comment by Nurdok** [comment]
No problem about the lag - I appreciate the review and it's not urgent.

* Regarding test location - I did this sort of on instinct. I guess putting the test next to `pydocstyle` would be more convenient. I'll go ahead and do that.
* Regarding `main` - I actually didn't know about `__main__.py`. I'll read about it and change it to support `python -m pydocstyle *`.

Thanks!

File: src/pydocstyle/cli.py (hunk 6 above)
Comment (Nurdok) [inline src/pydocstyle/cli.py]: Since this is a change in behavior, I'd rather do this in a separate PR from this refactoring.

**Comment by sigmavirus24** [comment]
@Nurdok having the tests in the package is perfectly fine too. I didn't mean to convince you to undo that. I'm just curious as to why you did it.

The reason I ask is because I have seen the benefit of always importing the fully-qualified module after it's been installed in the virtualenv via tox. This is why Flake8 has `src/flake8` and `tests/` so that the tests are easily editable (because I'm super lazy) and it *relies* on `flake8` being installed prior to running the tests. The layout is **personal preference.**

The reason I bring this up is because I think I saw relative imports in the tests and that concerns me a bit. If you're not testing what gets installed into a virtualenv, you're not testing what a user will receive after running `pip install pydocstyle`. Does that make sense?

File: src/pydocstyle/cli.py (hunk 6 above)
Comment (sigmavirus24) [inline src/pydocstyle/cli.py]: That sounds perfectly reasonable to me :)

**Comment by Nurdok** [comment]
I realize the test location is a preference, but as I read your suggestion it immediately became clear to me that it would solve all the ugly relative imports, so I'll take it :)

**Comment by sigmavirus24** [comment]
Hah, cool. =D

**Comment by Nurdok** [comment]
Do you see a reason not to put tests in `src`? i.e., have `pydocstyle` and tests siblings in the `src` directory. It will still have the benefits you describe, since the `tests` directory will not be installed in the virtualenv.

**Comment by sigmavirus24** [comment]
Nope. =)

**Comment by Nurdok** [comment]
I completed both points, although `__main__` does not work for Python 2.6.

**Comment by Nurdok** [comment]
@sigmavirus24 I merged this so that other work on pydocstyle can continue. If you have further comments, I'll be happy to hear them.

Cheers!
